### ISSUE TITLE
[R-package] CMake errors in MinGW builds on Windows

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#3060]
See [the most recent build on `master`](https://ci.appveyor.com/project/guolinke/lightgbm/builds/32728411/job/jpvhc1pkgghirltt). There are some errors being generated when building the R package using MinGW. This isn't causing any user-facing issues as far as I know, since it doesn't cause installation to fail, but we should fix these warnings so that they don't waste users' time or lead to unnecessary bug reports.

```text
installing via 'install.libs.R' to C:/projects/lightgbm/lightgbm.Rcheck/00LOCK-lightgbm/00new/lightgbm
[1] "Trying to build with MinGW"
CMake Error at C:/Program Files (x86)/CMake/share/cmake-3.16/Modules/CMakeMinGWFindMake.cmake:12 (message):
...
...
...
CMake Error: CMAKE_C_COMPILER not set, after EnableLanguage
CMake Error: CMAKE_CXX_COMPILER not set, after EnableLanguage
```

[This StackOverflow post](https://stackoverflow.com/a/32922266/3986677) suggested that maybe putting MinGW on `PATH` can fix this but so far I haven't had luck with it.

**Comment by jameslamb** [comment]
I just tried removing the extra mingw on PATH, per https://github.com/microsoft/LightGBM/pull/2965#discussion_r422487281, and I don't think that solved it.

You can see the changes  I made at https://github.com/jameslamb/LightGBM/pull/23/files and the corresponding build in AppVeyor at https://ci.appveyor.com/project/jameslamb/lightgbm/builds/32764241/job/cve3wk1f8obtw6q5

![image](https://user-images.githubusercontent.com/7608904/81484644-c58f7f00-920c-11ea-9a8d-b92b77caa43f.png)

**Comment by jameslamb** [comment]
I found one [StackOverflow answer](https://stackoverflow.com/a/32922266) suggesting that the message can be suppressed by installing `mingw32-base`.

Just to test, I installed [mingw-get-setup.exe](http://www.mingw.org/wiki/getting_started) and used it to install `mingw-base`. Even after doing that and adding `C:\MinGW\bin` to `PATH` (and restarting my shell), the error is still there.

**Comment by Laurae2** [comment]
cmake has a one time stopper if it finds any sh.exe in PATH, specifically it has an hardcoded MinGW one time stopper in cmake (intentional known issue on cmake end).

To circumvent it, cmake has to be ran twice. It keeps in cache the previous cmake run hence should not introduce major slowdown, however for users it will print a lot of unnecessary text and an unnecessary cmake error.

There is currently no workaround other than having a very recent cmake version. See the following cmake commit which lifts the one time stopper: https://gitlab.kitware.com/cmake/cmake/-/commit/82ddcf0db1d220564145122c3cce25d25ee0e254 (Nov 2, 2019).

(I could document better the workaround/reason here: https://github.com/microsoft/LightGBM/commit/3421bc6ce6f7294e22db7abad48a42e94fcc92cc#diff-473c89bfc9641f554ac12d37ebc3d377R61) + link to https://github.com/microsoft/LightGBM/pull/664#issuecomment-313464314

**Comment by jameslamb** [comment]
Thanks very much for the extra context! I saw comments in the code about running `cmake -G` twice for MinGW, but I didn't make the connection between that and the error messages mentioned in this issue.

I think we could close this issue just by adding a note to the R FAQs telling people they can safely ignore the errors https://lightgbm.readthedocs.io/en/latest/FAQ.html#r-package

**Comment by StrikerRUS** [comment]
Closed via #3078.
